pkg/applyconfigurations/gen.go

```diff
@@ -0,0 +1,256 @@
+	}
+	return isCRD(info)
+}
+
+// isCRD marks whether the type is a CRD based on the +kubebuilder:resource marker.
+func isCRD(info *markers.TypeInfo) bool {
+	objectEnabled := info.Markers.Get(isCRDMarker.Name)
+	return objectEnabled != nil
+}
+
+func (d Generator) Generate(ctx *genall.GenerationContext) error {
+	headerFilePath := d.HeaderFile
+
+	if headerFilePath == "" {
+		tmpFile, err := os.CreateTemp("", "applyconfig-header-*.txt")
+		if err != nil {
+			return fmt.Errorf("failed to create temporary file: %w", err)
+		}
+		tmpFile.Close()
+
+		defer os.Remove(tmpFile.Name())
+
+		headerFilePath = tmpFile.Name()
+	}
+
+	objGenCtx := ObjectGenCtx{
+		Collector:      ctx.Collector,
+		Checker:        ctx.Checker,
+		HeaderFilePath: headerFilePath,
+	}
+
+	for _, pkg := range ctx.Roots {
+		if err := objGenCtx.generateForPackage(pkg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ObjectGenCtx contains the common info for generating apply configuration implementations.
+// It mostly exists so that generating for a package can be easily tested without
+// requiring a full set of output rules, etc.
+type ObjectGenCtx struct {
+	Collector      *markers.Collector
+	Checker        *loader.TypeChecker
+	HeaderFilePath string
+}
+
+type Universe struct {
+	typeMetadata map[types.Type]*typeMetadata
+}
+
+type typeMetadata struct {
+	used bool
+}
+
+func (u *Universe) existingApplyConfigPath(_ *types.Named, pkgPath string) (string, bool) {
+	for prefix, replacePath := range importMapping {
+		if strings.HasPrefix(pkgPath, prefix) {
+			path := replacePath + strings.TrimPrefix(pkgPath, prefix)
+			return path, true
+		}
+	}
+	return "", false
+}
+
+func (u *Universe) IsApplyConfigGenerated(typeInfo *types.Named) bool {
+	if t, ok := u.typeMetadata[typeInfo]; ok {
+		return t.used
+	}
+	return false
+}
+
+func (u *Universe) GetApplyConfigPath(typeInfo *types.Named, pkgPath string) (string, bool) {
+	isApplyConfigGenerated := u.IsApplyConfigGenerated(typeInfo)
+	if path, ok := u.existingApplyConfigPath(typeInfo, pkgPath); ok {
+		if isApplyConfigGenerated {
+			return path, true
+		}
+		return pkgPath, false
+	}
+	// ApplyConfig is necessary but location is not explicitly specified. Assume the ApplyConfig exists at the below directory
+	if isApplyConfigGenerated {
+		return pkgPath + "/" + importPathSuffix, true
+	}
+	return pkgPath, false
+}
+
+// generateForPackage generates apply configuration implementations for
+// types in the given package, writing the formatted result to given writer.
+// May return nil if source could not be generated.
+func (ctx *ObjectGenCtx) generateForPackage(root *loader.Package) error {
+	enabled, _ := enabledOnPackage(ctx.Collector, root)
+	if !enabled {
+		return nil
+	}
+
+	genericArgs, _ := generatorargs.NewDefaults()
+	genericArgs.InputDirs = []string{root.PkgPath}
+	genericArgs.OutputPackagePath = filepath.Join(root.PkgPath, importPathSuffix)
+	genericArgs.GoHeaderFilePath = ctx.HeaderFilePath
```

Comment:
If I see correctly this currently stores the generated files somewhere below GOPATH/src. Would be probably good to align this to how we can set output dir in other generators.

Reply:
I tried to run this in Cluster API with ~ `apply paths=./api/...` I end up with files generated in $GOPATH/src/sigs.k8s.io/cluster-api/api/v1beta1/applyconfiguration/api/v1beta1

Reply:
Not sure if it's good if we store the applyconfigurations in a sub-package of the API types. Might be a simple output:apply:dir would be better and then we can store it below there, e.g. 
* output:apply:dir is explicitly set to `GOPATH/src/sigs.k8s.io/cluster-api/applyconfigurations`
* files are then generated in e.g. `$GOPATH/src/sigs.k8s.io/cluster-api/applyconfiguration/api/v1beta1`

I think this would follow the same pattern as the ones in client-go. (side note: the API folders are not well structured in Cluster API)

Reply:
(probably this is covered by: https://github.com/kubernetes-sigs/controller-tools/pull/818#discussion_r1199118670 feel free to resolve)